sway-lsp/src/capabilities/document_symbol.rs

```diff
@@ -1,60 +1,499 @@
+                    let children = collect_struct_fields(&struct_decl);
+                    let struct_symbol = DocumentSymbolBuilder::new()
+                        .name(span.str().to_string())
+                        .kind(lsp_types::SymbolKind::STRUCT)
+                        .range(range)
+                        .selection_range(range)
+                        .children(children)
+                        .build();
+                    Some(struct_symbol)
+                }
+                TyDecl::AbiDecl(decl) => {
+                    let abi_decl = engines.de().get_abi(&decl.decl_id);
+                    let decl_str = abi_decl.span().str();
+                    let name = extract_header(&decl_str);
+                    let range = get_range_from_span(&abi_decl.name.span());
+                    let children = collect_fns_from_abi_decl(engines, &abi_decl);
+                    let abi_symbol = DocumentSymbolBuilder::new()
+                        .name(name)
+                        .kind(lsp_types::SymbolKind::NAMESPACE)
+                        .range(range)
+                        .selection_range(range)
+                        .children(children)
+                        .build();
+                    Some(abi_symbol)
+                }
+                TyDecl::TraitDecl(decl) => {
+                    let trait_decl = engines.de().get_trait(&decl.decl_id);
+                    let decl_str = trait_decl.span().str().to_string();
+                    let name = extract_header(&decl_str);
+                    let range = get_range_from_span(&trait_decl.name.span());
+                    let children =
+                        collect_interface_surface(engines, &trait_decl.interface_surface);
+                    let trait_symbol = DocumentSymbolBuilder::new()
+                        .name(name)
+                        .kind(lsp_types::SymbolKind::INTERFACE)
+                        .range(range)
+                        .selection_range(range)
+                        .children(children)
+                        .build();
+                    Some(trait_symbol)
+                }
+                TyDecl::TraitTypeDecl(decl) => {
+                    let trait_type_decl = engines.de().get_type(&decl.decl_id);
+                    Some(build_trait_symbol(&trait_type_decl))
+                }
+                TyDecl::ImplSelfOrTrait(decl) => {
+                    let impl_trait_decl = engines.de().get_impl_self_or_trait(&decl.decl_id);
+                    let decl_str = impl_trait_decl.span().str().to_string();
+                    let name = extract_header(&decl_str);
+                    let range = get_range_from_span(&impl_trait_decl.trait_name.suffix.span());
+                    let children = collect_ty_trait_items(engines, &impl_trait_decl.items);
+                    let symbol = DocumentSymbolBuilder::new()
+                        .name(name)
+                        .kind(lsp_types::SymbolKind::NAMESPACE)
```

Comment:
Just checking if this is supposed to be `NAMESPACE` or not.

Reply:
I just followed what rust-analyzer uses for traits so we have the same look. In this case `lsp_types::SymbolKind::NAMESPACE` is used.